Subcell: Add action TciAndSwitchToDg

nilsdeppe opened the pull request:
## Proposed changes

Add an action that runs the TCI on the subcell grid and then decides whether or not to switch back to DG. The decision making process is a bit tedious because we ultimately need to support substep, multistep, one high-order one-step (ADER) methods

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add action TciAndSwitchToDg

fmahebert commented on src/Evolution/DgSubcell/Actions/TciAndSwitchToDg.hpp:
```diff
@@ -0,0 +1,290 @@
+ * 1. if we are using a substep time integrator and are not at the end of a
+ *    step, or we are in the self-starting stage of a multistep method, or the
+ *    `subcell_options.always_use_subcells() == true`, then we do not run any
+ *    TCI or try to go back to DG. We need to avoid reconstructing the time
```
By "reconstructing the time stepper history", I assume you mean reconstructing the DG soln from the subcell values? Do you think it would be useful to clarify this (e.g., like in item 2 below)? I'm just thinking that "reconstruction" can also be used as a FD-only operation, which isn't really the one you're performing here I think?

fmahebert commented on src/Evolution/DgSubcell/Actions/TciAndSwitchToDg.hpp:
```diff
@@ -0,0 +1,290 @@
+ *    to subcell.
+ * 2. Reconstruct the subcell solution to the DG grid.
+ * 3. Run the relaxed discrete maximum principle (RDMP) troubled-cell indicator
+ *    (TCI), checking both the subcell solution at \f$t^{n+1}\f$ and the
```
t^{n+1} meaning that you're checking the "candidate" solution for the next time step before considering the time step to be complete? It would be nice to discuss somewhere (perhaps before this list of steps in the algorithm) where the action fits into the DG-subcell scheme.

fmahebert commented on src/Evolution/DgSubcell/Actions/TciAndSwitchToDg.hpp:
```diff
@@ -0,0 +1,290 @@
+ *    (TCI), checking both the subcell solution at \f$t^{n+1}\f$ and the
+ *    reconstructed DG solution at \f$t^{n+1}\f$ to make sure they are
+ *    admissible.
+S * 4. If the RDMP TCI marked the DG solution as admissible, run the
```
Sremove the extra 'S' at the beginning of this line

fmahebert commented on src/Evolution/DgSubcell/Actions/TciAndSwitchToDg.hpp:
```diff
@@ -0,0 +1,290 @@
+
+    const TimeStepId& time_step_id = db::get<::Tags::TimeStepId>(box);
+    const SubcellOptions& subcell_options = db::get<Tags::SubcellOptions>(box);
+    if (db::get<::Tags::TimeStepId>(box).substep() != 0 or
```
Why not `time_step_id.substep()` since you've already assigned a variable for this?

fmahebert commented on src/Evolution/DgSubcell/Actions/TciAndSwitchToDg.hpp:
```diff
@@ -0,0 +1,290 @@
+            using std::swap;
+            swap(*active_vars_ptr, *inactive_vars_ptr);
+
+            // Project the time stepper history to the subcells.
```
Isn't this backwards? Based on item 5 from the main docs above and the code below, I was expecting this to say something like "Reconstruct the DG solution for each time in the time stepper history"

nilsdeppe commented:
I've rebased and pushed a fixup. Thanks for the review!

fmahebert reviewed (commented):
LGTM, please squash

nilsdeppe commented:
Thanks for looking at the fixup so quickly! I've squashed that in :)

nilsdeppe commented:
I've rebased this on develop and also added a fixup that resets the `DidRollback` to `false` in this action. The change is needed to avoid unintentionally recomputing primitives when they aren't needed.

fmahebert commented:
Seems good. But can you add the action's action on `DidRollback` to the documentation? I'm thinking at least in the list of Databox quantities modified, and perhaps also in words above (up to you).

nilsdeppe commented:
Agreed! Pushed a fixup that adds some documentation on that front. Please let me know if you have any suggestions for improving the documentation :)

fmahebert commented:
No further comments, the doc fixup looks good! Please squash.

nilsdeppe commented:
Rebased and squashed. Thanks for reviewing :)

kidder commented on src/Evolution/DgSubcell/Actions/TciAndSwitchToDg.hpp:
```diff
@@ -0,0 +1,308 @@
+ * grid.
+ *
+ * The algorithm proceeds as follows:
+ * 1. if we are using a substep time integrator and are not at the end of a
```
if -> If

kidder commented on tests/Unit/Evolution/DgSubcell/Actions/Test_TciAndSwitchToDg.cpp:
```diff
@@ -0,0 +1,386 @@
+  // 3. Check if always_use_subcells, inactive vars are untouched.
+  // 4. Check if RDMP gets triggered, then tci_mutator is not called, and we
+  //    stay on subcell
+  // 5. Check if RDMP is not triggered, but tci_mutator is, we stay SCL
```
SCL -> on subcell

kidder reviewed (changes requested):
you can squash immediately

nilsdeppe commented:
I've rebased (conflict in  CMakeLists.txt) and squashed. Thanks for the reviews :)
